File: Version1/Templates/MVC + UITableView.xctemplate/___FILEBASENAME___/Controller/___FILEBASENAME___ViewController.m
//
//  ___FILENAME___
//  ___PROJECTNAME___
//
//  Created by ___FULLUSERNAME___ on ___DATE___.
//___COPYRIGHT___
//

#import "___VARIABLE_moduleName___ViewController.h"
#import "___VARIABLE_moduleName___Model.h"
#import "___VARIABLE_moduleName___View.h"


@interface ___VARIABLE_moduleName___ViewController () <___VARIABLE_moduleName___ModelDelegate, ___VARIABLE_moduleName___ViewDelegate, UITableViewDelegate, ___VARIABLE_moduleName___CellDelegate>

@property (nonatomic, strong) NSObject<___VARIABLE_moduleName___ModelProtocol> *model;

/** @brief Overriding view property */
@property (nonatomic, strong) UIView<___VARIABLE_moduleName___ViewProtocol> *view;

@property (nonatomic, strong) ___VARIABLE_moduleName___DataSource * dataSource;

@property (nonatomic, strong) ___VARIABLE_moduleName___Router *router;

@end


@implementation ___VARIABLE_moduleName___ViewController

@dynamic view;
@dynamic model;
@dynamic router;


#pragma mark - Initialization

- (nullable id)initWithView:(nonnull UIView<___VARIABLE_moduleName___ViewProtocol> *)view model:(nonnull NSObject<___VARIABLE_moduleName___ModelProtocol> *)model router:(nullable ___VARIABLE_moduleName___Router *)router dataSource:(nonnull ___VARIABLE_moduleName___DataSource *)dataSource {
    self = [super initWithNibName:NSStringFromClass([self class]) bundle:nil view:view model:model];
    if (self) {
        self.model.delegate = self;
        self.router = router;
        
        _dataSource = dataSource;
        _dataSource.cellDelegate = self;
        
        
        // your custom code
    }
    return self;
}


#pragma mark - View lifecycle

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.view.delegate = self;
    [self connectTableViewDependencies];
    
    // your custom code
}

- (void)connectTableViewDependencies {
    self.view.tableView.delegate = self;
    [self.dataSource registerNibsForTableView:self.view.tableView];
    self.view.tableView.dataSource = self.dataSource;
}

#pragma mark - UITableView delegate

/** Implement UITableViewDelegate methods here */

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    [self.router navigateToSomeScreenFromViewController:self withBackgroundColor:[UIColor orangeColor]];
}


#pragma mark - ___VARIABLE_moduleName___ViewDelegate

/** Implement ___VARIABLE_moduleName___ViewDelegate methods here */

- (void)viewSomeAction:(UIView<___VARIABLE_moduleName___ViewProtocol> *)view {
    NSLog(@"Something happened on view!");
}


#pragma mark - ___VARIABLE_moduleName___ModelDelegate

- (void)modelDidChanged:(NSObject<___VARIABLE_moduleName___ModelProtocol> *)model {
    
}


#pragma mark - ___VARIABLE_moduleName___CellDelegate

/** Implement ___VARIABLE_moduleName___CellDelegate methods here */

- (void)cellDidTapSomeButton:(UITableViewCell *)cell {
    
    NSIndexPath *indexPath = [self.view.tableView indexPathForCell:cell];
    
    NSLog(@"Cell at index path (section: %ld, row: %ld) did tap some button", (long)indexPath.section, (long)indexPath.row);
}


#pragma mark - Private

/** Implement any private methods here */

@end

File: Version1/Templates/MVC Swift + UITableView.xctemplate/___FILEBASENAME___/___FILEBASENAME___ViewController.swift
//
//  ___FILENAME___
//  ___PROJECTNAME___
//
//  Created by ___FULLUSERNAME___ on ___DATE___.
//  ___COPYRIGHT___
//

import UIKit

typealias ___VARIABLE_moduleName___ViewControllerType = MVCViewController<___VARIABLE_moduleName___ModelProtocol, ___VARIABLE_moduleName___ViewProtocol, ___VARIABLE_moduleName___Router>

class ___VARIABLE_moduleName___ViewController: ___VARIABLE_moduleName___ViewControllerType, UITableViewDelegate {

    private var dataSource: ___VARIABLE_moduleName___DataSource!

    // MARK: - Initializers

    convenience init(withView view: ___VARIABLE_moduleName___ViewProtocol, model: ___VARIABLE_moduleName___ModelProtocol, router: ___VARIABLE_moduleName___Router, dataSource: ___VARIABLE_moduleName___DataSource) {

        self.init(withView: view, model: model, router: router)

        self.model.delegate = self

        self.dataSource = dataSource
        self.dataSource.cellDelegate = self

        // your custom code
    }

    required init(withView view: ___VARIABLE_moduleName___ViewProtocol!, model: ___VARIABLE_moduleName___ModelProtocol!, router: ___VARIABLE_moduleName___Router?) {
        super.init(withView: view, model: model, router: router)
    }

    // MARK: - View life cycle

    override func viewDidLoad() {
        super.viewDidLoad()

        self.customView.delegate = self
        self.connectTableViewDependencies()
    }

    private func connectTableViewDependencies() {

        self.customView.tableView.delegate = self
        self.dataSource.registerNibsForTableView(tableView: self.customView.tableView)
        self.customView.tableView.dataSource = self.dataSource
    }

    // MARK: - Table view delegate

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {

        tableView.deselectRow(at: indexPath, animated: true)

        self.router?.navigateToSomeScreen(from: self, withBackgroundColor: UIColor.gray)
    }
}

// MARK: - ___VARIABLE_moduleName___ViewDelegate

extension ___VARIABLE_moduleName___ViewController: ___VARIABLE_moduleName___ViewDelegate {

    func viewSomeAction(view: ___VARIABLE_moduleName___ViewProtocol) {
    }
}

// MARK: - ___VARIABLE_moduleName___ModelDelegate

extension ___VARIABLE_moduleName___ViewController: ___VARIABLE_moduleName___ModelDelegate {

    func modelDidChanged(model: ___VARIABLE_moduleName___ModelProtocol) {
    }
}

// MARK: - ___VARIABLE_moduleName___CellDelegate

extension ___VARIABLE_moduleName___ViewController: ___VARIABLE_moduleName___CellDelegate {

    func cellDidTapSomeButton(cell: ___VARIABLE_moduleName___TableViewCell) {
    }
}
